llvm/lib/Target/AMDGPU/SOPInstructions.td

```diff
@@ -1865,6 +1870,12 @@ let SubtargetPredicate = isNotGFX9Plus in {
 def : GetFPModePat<fpmode_mask_gfx6plus>;
 }
 
+let SubtargetPredicate = isGFX9GFX10 in
+def : GCNPat<
+  (int_amdgcn_pops_exiting_wave_id),
+  (S_MOV_B32_sideeffects (i32 SRC_POPS_EXITING_WAVE_ID))
```

Comment:
Does it really need side effects? If it's directly reading a reserved register you can't do much of anything with it anyway

Reply:
It needs side effects to prevent tablegen from complaining about selecting a side-effecting intrinsic to a non-side-effecting instruction.

Reply:
InaccessibleMemOnly should avoid that

Reply:
No. It doesn't need a memory operand if it's a register dependency

Reply:
Then what would stop the MachineInstr getting CSEd, hoisted out of loops, rematerialized etc? We definitely need to prevent it from being moved past control flow so that you can write loops that poll the value.

Reply:
The use of reserved physical register

Reply:
Ah, I did not know it had that effect. Thanks.

Reply:
IntrReadMem+InaccessibleMemOnly? The TableGen enforcement of this stuff has always been broken

Reply:
I have changed it to use InaccessibleMemOnly. I don't think I can use IntrReadMem because that would allow CSEing these in IR, wouldn't it?

Reply:
No. The reason this is so difficult is the tablegen annotations for intrinsics are backwards from how the attributes work. IntrReadMem really means does not write memory

Reply:
"No" to which part? I have tried adding IntrReadMem and it does allow the intrinsic calls to be CSE'd so I do not want that.

Reply:
This sounds like a bug to me. I would assume inaccessiblememonly + readnone is not sufficient to allow CSE

Reply:
It isn't CSE'd with a possibly writing call in between. Is this supposed to act more like readcyclecounter?

Reply:
> Is this supposed to act more like readcyclecounter?

Yes. It's a volatile value that can change at any time, even without an explicit write in the IR.

Reply:
I guess you just have to copy the attributes from readyclecounter, which is just DefaultAttrsIntrinsic

Reply:
That's kind of what I've done already! readcyclecounter reads and writes mem. The new instrinsic also reads and writes mem. The only difference is that I have restricted it to inaccessible mem only, which I still think is reasonable (and could perhaps be applied to readcyclecounter too).

The current patch works for me. Are you asking for / suggesting any changes?

Reply:
> > Then what would stop the MachineInstr getting CSEd, hoisted out of loops, rematerialized etc? We definitely need to prevent it from being moved past control flow so that you can write loops that poll the value.
>
> The use of reserved physical register

That does not seem to work. See: https://github.com/llvm/llvm-project/blob/af82e63c28f67bf61a9b2b0e64bc55be4acf520e/llvm/test/CodeGen/AMDGPU/llvm.amdgcn.pops.exiting.wave.id.ll#L39 where `%1:sreg_32 = S_MOV_B32 $src_pops_exiting_wave_id` is hoisted by early-machinelicm.